test/datatools/test.cc

```diff
@@ -0,0 +1,249 @@
+
+    // Read every speed tile, and update it with fixed speed of `new_speed` km/h (original speeds are
+    // 10km/h)
+    valhalla::baldr::GraphReader reader(config.get_child("mjolnir"));
+    mtar_header_t tar_header;
+    while ((mtar_read_header(&tar, &tar_header)) != MTAR_ENULLRECORD) {
+      valhalla::baldr::TrafficTile tile(
+          std::make_unique<MMapGraphMemory>(memory,
+                                            reinterpret_cast<char*>(tar.stream) + tar.pos +
+                                                sizeof(mtar_raw_header_t_),
+                                            tar_header.size));
+
+      valhalla::baldr::GraphId tile_id(tile.header->tile_id);
+
+      for (uint32_t index = 0; index < tile.header->directed_edge_count; index++) {
+        auto* current = const_cast<valhalla::baldr::TrafficSpeed*>(tile.speeds + index);
+        setter_cb(reader, tile, index, current);
+      }
+      mtar_next(&tar);
+    }
+  }
+}
+
+void customize_historical_traffic(const boost::property_tree::ptree& config,
+                                  const HistoricalTrafficCustomize& cb) {
+  // loop over all tiles in the tileset
+  valhalla::baldr::GraphReader reader(config.get_child("mjolnir"));
+  auto tile_dir = config.get<std::string>("mjolnir.tile_dir");
+  for (const auto& tile_id : reader.GetTileSet()) {
+    valhalla::mjolnir::GraphTileBuilder tile(tile_dir, tile_id, false);
+    std::vector<valhalla::baldr::DirectedEdge> edges;
+    edges.reserve(tile.header()->directededgecount());
+    for (const auto& edge : tile.GetDirectedEdges()) {
+      edges.push_back(edge);
+      const auto historical = cb(edges.back());
+      if (historical) {
+        auto coefs = valhalla::baldr::compress_speed_buckets(historical->data());
+        tile.AddPredictedSpeed(edges.size() - 1, coefs, tile.header()->directededgecount());
+      }
+      edges.back().set_has_predicted_speed(static_cast<bool>(historical));
+    }
+    tile.UpdatePredictedSpeeds(edges);
+  }
+}
+
+void customize_edges(const boost::property_tree::ptree& config, const EdgesCustomize& setter_cb) {
+  // loop over all tiles in the tileset
+  valhalla::baldr::GraphReader reader(config.get_child("mjolnir"));
+  auto tile_dir = config.get<std::string>("mjolnir.tile_dir");
+  for (const auto& tile_id : reader.GetTileSet()) {
+    valhalla::mjolnir::GraphTileBuilder tile(tile_dir, tile_id, false);
+    std::vector<valhalla::baldr::NodeInfo> nodes;
+    nodes.reserve(tile.header()->nodecount());
+    for (const auto& node : tile.GetNodes())
+      nodes.push_back(node);
+
+    std::vector<valhalla::baldr::DirectedEdge> edges;
+    edges.reserve(tile.header()->directededgecount());
+
+    GraphId edgeid = tile_id;
+    for (size_t j = 0; j < tile.header()->directededgecount(); ++j, ++edgeid) {
+      edges.push_back(tile.directededge(j));
+      setter_cb(edgeid, edges.back());
+    }
+    tile.Update(nodes, edges);
+  }
+}
```

Comment:
these two functions look like the only two things that wouldnt work if a person build the repository with datatools turned off. which even more makes me think it would be fine to just ifdef them rather than make these huge changes.

did i miss something here???

Reply:
Hm, maybe you are right. I need to use `add_compile_definitions` to add define, right?